Convert 'Remote access API' wiki page to jenkins.io #3184 (#3262)
jenkins-infra/jenkins.io: content/doc/book/using/remote-access-api.adoc

authenticate remote API requests. See
https://wiki.jenkins.io/display/JENKINS/Authenticating+scripted+clients[Authenticating
scripted clients] for more details.

[[RemoteaccessAPI-CSRFProtection]]
== CSRF Protection

*Note*: API tokens are preferred instead of crumbs for CSRF protection.

[[RemoteaccessAPI-XPathselection]]
== XPath selection

The XML API supports a selection by XPath by using the query parameter
'xpath'. This is convenient for extracting information in environments
where XML manipulation is tedious (such as shell script.) See
https://issues.jenkins-ci.org/browse/JENKINS-626[issue #626] for an
example of how to use this. +
See `+.../api/+` on your Jenkins server for more up-to-date details.

[[RemoteaccessAPI-XPathexclusion]]
=== XPath exclusion

Similar to the 'xpath' query parameter above, you can use (possibly
multiple) 'exclude' query patterns to exclude nodes from the resulting
XML. All the nodes that match the specified XPath will be removed from
the XML. +
See `+.../api/+` on your Jenkins server for more up-to-date details.

[[RemoteaccessAPI-Depthcontrol]]
== Depth control

Sometimes the remote API doesn't give you enough information in one

> call. For example, if you'd like to find out all the last successful

Reviewer: Remove some extra words
```suggestion
call. For example, if you'd like to find the last successful
```

build of a given view, you'd realize that the invocation to the remote
API of the view won't give you this, and you'd have to recursively call

> the remote API of each project to find this out. The depth control,
> solves this problem. To understand this feature,
> it's good to start with how the remote API works.

Reviewer: ```suggestion
the remote API of each project.
Depth control solves this problem. 
Depth control is fundamentally connected to the Jenkins data model.
```

The data model that Jenkins maintains internally can be thought of as a
big tree structure, and when you make a remote API call, you are getting
a small subtree of it. The subtree is rooted at the object for which you
made a remote API call, and the sub-tree is cut beyond certain depth to
avoid returning too much data. You can adjust this cut-off behavior by
specifying the depth query parameter. When you specify a positive depth
value, the subtree cut-off happens that much later.

So the net result is, if you specify a bigger depth value, you'll see
that the remote API will now return more data. Because of the algorithm,
this works in such a way that the data returned by a bigger depth value
includes all the data returned by smaller depth value.

See `+.../api/+` on your Jenkins server for more up-to-date details.

[[RemoteaccessAPI-PythonAPIwrappers]]
== Python API wrappers

https://pypi.python.org/pypi/jenkinsapi[JenkinsAPI] and
https://pypi.python.org/pypi/python-jenkins/[Python-Jenkins] are

> object-oriented python wrappers for the Python REST API which aims to

Reviewer: Replace singular with plural
```suggestion
object-oriented python wrappers for the Python REST API which aim to
```

provide a more conventionally pythonic way of controlling a Jenkins
server. It provides a higher-level API containing a number of
convenience functions. Services offered currently include:

* Query the test-results of a completed build
* Get objects representing the latest builds of a job
* Search for artifacts by simple criteria
* Block until jobs are complete
* Install artifacts to custom-specified directory structures

> * username/password auth support for jenkins instances with auth turned
> on

Reviewer: ```suggestion
* Authentication support for Jenkins instances
```

* Ability to search for builds by subversion revision

> * Ability to add/remove/query jenkins slaves

Reviewer: Agents rather than slaves is the preferred word
```suggestion
* Ability to add/remove/query Jenkins agents
```

[[RemoteaccessAPI-RubyAPIwrappers]]
== Ruby API wrappers

https://rubygems.org/gems/jenkins_api_client[Jenkins API Client] is an
object oriented ruby wrapper project that consumes Jenkins's JSON API
and aims at providing access to all remote API Jenkins provides. It is

> available as a Rubygem and can be useful to interact with the Job, Node,

Reviewer: ```suggestion
available as a Ruby gem and can be useful to interact with the Job, Node,
```

View, BuildQueue, and System related functionalities. Services currently
offered include:

* Creating jobs by sending xml file or by specifying params as options
with more customization options including source control, notifications,
etc.
* Building jobs (with params), stopping builds, querying details of
recent builds, obtaining build params, etc.
* Listing jobs available in Jenkins with job name filter, job status
filter.
* Adding/removing downstream projects.
* Chaining jobs i.e given a list of projects each project is added as a
downstream project to the previous one.
* Obtaining progressive console output.
* Username/password based authentication.
* Command Line Interface with a lot of options provided in the
libraries.
* Creating, listing views.
* Adding jobs to views and removing jobs from views.

> * Adding/removing jenkins slaves, querying details of slaves.

Reviewer: ```suggestion
* Adding/removing Jenkins agents, querying details of agents.
```

* Obtaining the tasks in build queue, and their age, cause, reason, ETA,
ID, params and much more.
* Quiet down, cancel quiet down, safe restart, force restart, and wait
till Jenkins becomes available after a restart.
* Ability to list installed/available plugins, obtain information about
plugins, install/uninstall plugins and much more with plugins.

> This project is in development and new features are getting added over time.
> Watch the progress on new features

Reviewer: Last commit was in 2018.
```suggestion
The project source code is at
```

https://github.com/arangamani/jenkins_api_client[here].

[[RemoteaccessAPI-JavaAPIwrappers]]
== Java API wrappers

The https://github.com/cdancy/jenkins-rest[jenkins-rest] library is an

> object oriented java project that provides access to the Jenkins REST

Reviewer: ```suggestion
object oriented Java project that provides access to the Jenkins REST
```

> API programmatically to some remote API Jenkins provides. It is build

Reviewer: ```suggestion
API programmatically to some remote API Jenkins provides. It is built
```

> using the amazing https://jclouds.apache.org/[jclouds toolkit] and can

Reviewer: ```suggestion
using the https://jclouds.apache.org/[jclouds toolkit] and can
```

easily be extended to support more REST endpoints. Its feature set
evolves and users are invited to contribute new endpoints via
pull-requests. In its current state it is possible with this library to

> submit a job, track its progress through the queue, and during its

Reviewer: ```suggestion
submit a job, track its progress through the queue, and monitor its
```

execution until its completion, and obtain the build status. Services
currently offered include:

* Endpoint definition (property or environment variable)
* Authentication (basic and API token via property or environment
variable)
* Crumbs Issuer support (auto-detect crumbs)
* Folder support
* Jobs API (build, buildInfo, buildWithParameters, config, create,
delete, description, disable, enable, jobInfo, lastBuildNumber,
lastBuidTimestamp and progressiveText)
* Plugin manager API (installNecessaryPlugins, list current plugins)
* Queue API (cancel, list queue items, query queue item)
* Statistics API (overall load)
* Systems API (systemInfo)

The project can evolve rapidly, this list is accurate only as of the
date of writing.

[[RemoteaccessAPI-DetectingJenkinsversion]]
== Detecting Jenkins version

> To check the version of Jenkins, load the top page (or, as of 1.483, any
> `+.../api/*+` page too) and check for the `+X-Jenkins+` response header.

Reviewer: No need to disclaim a build that is more than 6 years old.
```suggestion
To check the version of Jenkins, load the top page or any
`+.../api/*+` page and check for the `+X-Jenkins+` response header.
```

---
layout: section
title: Remote Access API
---

Jenkins provides machine-consumable remote access API to its
functionalities. Currently it comes in three flavors:

. XML
. JSON with JSONP support
. Python

Remote access API is offered in a REST-like style. That is, there is no
single entry point for all features, and instead they are available
under the `+".../api/"+` URL where `+"..."+` portion is the data that it
acts on.

> For example, if your Jenkins installation sits at http://ci.jruby.org/,

Reviewer 2: I would recommend using https://ci.jenkins.io in the examples, this is long dead

Author: ok

Please note, in this example, the symbol '@' is important to mention. Also, the path to the file is absolute path.
In order to make this command work, you need to configure your Jenkins job to take a file parameter and 'name' in this command corresponds to 'file location' field in the Jenkins job configuration.

In above example, 'harness' is just a name of folder that may not exist in your workspace, you can write "name":"Task.xml" and it will place the Task.xml at root of your workspace.
Remember that name in this command should match with File location in file parameter in job configuration.
----

image:../images/using/remote-access-api/file-parameter.png[image,title="File Parameter"]

[[RemoteaccessAPI-RemoteAPIandsecurity]]
== Remote API and security

When your Jenkins is secured, you can use HTTP BASIC authentication to
authenticate remote API requests. See
https://wiki.jenkins.io/display/JENKINS/Authenticating+scripted+clients[Authenticating
scripted clients] for more details.

[[RemoteaccessAPI-CSRFProtection]]
== CSRF Protection

*Note*: From Jenkins 2.96 onward, you can use an API token and avoid
using a crumb / CSRF token.

> If your Jenkins uses the "Prevent Cross Site Request Forgery exploits"

Reviewer 2: you can no longer turn off csrf prevention now so this sentence isn't needed.

I'm not sure if crumb mentioning is needed at all,

API tokens should be used instead

security option (which it should), when you make a `+POST+` request, you
have to send a CSRF protection token as an HTTP request header. +
For curl/wget you can obtain the header needed in the request from the
URL `+JENKINS_URL/crumbIssuer/api/xml+` (or `+.../api/json+`). Something
like this:

....
wget -q --auth-no-challenge --user USERNAME --password PASSWORD --output-document - \
'JENKINS_URL/crumbIssuer/api/xml?xpath=concat(//crumbRequestField,":",//crumb)'
....

This will print something like "Jenkins-Crumb:1234abcd", which you
should add to the subsequent request.

> [[RemoteaccessAPI-Samplecode]]
> == Sample code
> 
> https://github.com/jenkinsci/extras-client-demo/tree/master/src/main/java/org/jvnet/hudson/client_demo/[A
> simple client is available] to demonstrate how you can invoke the XML
> from Java
> (https://github.com/jenkinsci/extras-client-demo/blob/master/src/main/java/org/jvnet/hudson/client_demo/Main.java[Java
> source])

Reviewer: This example code no longer compiles since it is based on Java 5.  I recommend that we remove it.

If we really need it, then it needs to have https://github.com/jenkinsci/extras-client-demo/pull/2 merged to make it compile again.
```suggestion
```

[[RemoteaccessAPI-XPathselection]]
== XPath selection

The XML API supports a selection by XPath by using the query parameter
'xpath'. This is convenient for extracting information in environments
where XML manipulation is tedious (such as shell script.) See
https://issues.jenkins-ci.org/browse/JENKINS-626[issue #626] for an
example of how to use this. +
See `+.../api/+` on your Jenkins server for more up-to-date details.

[[RemoteaccessAPI-XPathexclusion]]
=== XPath exclusion

Similar to the 'xpath' query parameter above, you can use (possibly
multiple) 'exclude' query patterns to exclude nodes from the resulting
XML. All the nodes that match the specified XPath will be removed from
the XML. +
See `+.../api/+` on your Jenkins server for more up-to-date details.

[[RemoteaccessAPI-Depthcontrol]]
== Depth control

Sometimes the remote API doesn't give you enough information in one
call. For example, if you'd like to find out all the last successful
build of a given view, you'd realize that the invocation to the remote
API of the view won't give you this, and you'd have to recursively call
the remote API of each project to find this out. The depth control,

> introduced in 1.167, solves this problem. To understand this feature,

Reviewer: No benefit to referencing a Jenkins version from 5+ years ago.  We can safely act like the capability has always been available.
```suggestion
solves this problem. To understand this feature,
```

it's good to start with how the remote API works.

The data model that Jenkins maintains internally can be thought of as a
big tree structure, and when you make a remote API call, you are getting
a small subtree of it. The subtree is rooted at the object for which you
made a remote API call, and the sub-tree is cut beyond certain depth to
avoid returning too much data. You can adjust this cut-off behavior by
specifying the depth query parameter. When you specify a positive depth
value, the subtree cut-off happens that much later.

So the net result is, if you specify a bigger depth value, you'll see
that the remote API will now return more data. Because of the algorithm,
this works in such a way that the data returned by a bigger depth value

> includes all the data returned by smaller +
> depth value.

Reviewer: No need for the line continuation character
```suggestion
includes all the data returned by smaller depth value.
```

See `+.../api/+` on your Jenkins server for more up-to-date details.

[[RemoteaccessAPI-PythonAPIwrappers]]
== Python API wrappers

> http://pypi.python.org/pypi/jenkinsapi[JenkinsAPI] and

Reviewer 2: ```suggestion
https://pypi.python.org/pypi/jenkinsapi[JenkinsAPI] and
```

offered include:

* Creating jobs by sending xml file or by specifying params as options
with more customization options including source control, notifications,
etc.
* Building jobs (with params), stopping builds, querying details of
recent builds, obtaining build params, etc.
* Listing jobs available in Jenkins with job name filter, job status
filter.
* Adding/removing downstream projects.
* Chaining jobs i.e given a list of projects each project is added as a
downstream project to the previous one.
* Obtaining progressive console output.
* Username/password based authentication.
* Command Line Interface with a lot of options provided in the
libraries.
* Creating, listing views.
* Adding jobs to views and removing jobs from views.
* Adding/removing jenkins slaves, querying details of slaves.
* Obtaining the tasks in build queue, and their age, cause, reason, ETA,
ID, params and much more.
* Quiet down, cancel quiet down, safe restart, force restart, and wait
till Jenkins becomes available after a restart.
* Ability to list installed/available plugins, obtain information about
plugins, install/uninstall plugins and much more with plugins.

> This project is in rapid development and new features are getting added

Reviewer 2: this sentence doesn't seem accurate (no commits in quite awhile)

under the `+".../api/"+` URL where `+"..."+` portion is the data that it
acts on.

For example, if your Jenkins installation sits at https://ci.jenkins.io,
visiting https://ci.jenkins.io/api/ will show just the top-level API
features available – primarily a listing of the configured jobs for this
Jenkins instance. +
Or if you want to access information about a particular build, e.g.

> https://ci.jenkins.io/job/jruby-base/lastSuccessfulBuild/, then go to

Reviewer 3: I think this should either be non-interactive links to a fictional server `https://ci.example.com` or working links to  a publicly available job, e.g.
https://ci.jenkins.io/job/Core/job/jenkins/job/master/lastSuccessfulBuild/
https://ci.jenkins.io/job/Core/job/jenkins/job/master/lastSuccessfulBuild/api

Author: Thanks @zbynek, I've replaced the links
